Evp kernel_v2

mhrib opened the pull request:
#205 updated to present master.
+ Added "evp_1d"+"evp_2d" timer
+ Removed  "strocnx,strocny" from stepu subroutine in dyn_evp_shared

-----

New vectorized EVP kernel:
- vectorized - contiguous vectors.
- Running OpenMP on the GLOBAL net on one single node ie. reduce communication and avoiding HALO updates to every single block every iteration.
- Able to run CICE mixed MPI+OpenMP on several nodes. Use Node "my_task==master_task" for EVP iterations:
- Works for "regular" circular grids - eg. GX3. Ie. do a "circular" halo update.
- Not tested for tri-pole grids. Doubt that it works. A stop is introduced in the code.
- Namelist: Shift between "CICE standard" EVP and this EVP via. namelist. Default is evp_kernel_ver=0 , which is "CICE standard" version
- Use environment "OMP_NUM_THREADS" to select number of threads used by the evp_kernel.
- There is a natural "overhead" associated with gather/scatter and defining vectors. Therefore it becomes more efficient for "large" setups.

Possible code correction/reduction??:  affects "stepu" routines in: evp_kernel1d.F90 *AND* default code ice_dyn_shared.F90:
- Is "strocnx,strocny" used at all from subroutine "stepu"? Is it needed here? It is also calculated in subroutine "dyn_finish" short after the EVP-loop. Therefore I have removed the calculation in this EVP_kernel, but it can easily be re-introduced (using slightly more memory) by searching for: "!strocn"

--o--

Namelist:
&dynamics_nml
evp_kernel_ver = 0  ! 0: CICE (default) , 2: kernel_v2

Environment:
OMP_NUM_THREADS

Option: REAL4 internally instead of REAL8:
- Not implemented as a namelist switch, but can be tried out simply by this "poor-man-option":
  mv evp_kernel1d.F90 evp_kernel1d_r8.F90
  cat evp_kernel1d_r8.F90 | sed s/DBL_KIND/REAL_KIND/g > evp_kernel1d.F90

--o--

- Developer(s):
Mads Hvid Ribergaard and Jacob Weismann Poulsen, DMI

- Please suggest code Pull Request reviewers in the column at right.
@eclare108213

- Are the code changes bit for bit, different at roundoff level, or more substantial?
In most cases it should be "bit-to-bit". Depending on how the code is translated during compilation and which math libs are used.

- Does this PR create or have dependencies on Icepack or any other models?
Nothing other than CICE normally have

- Is the documentation being updated with this PR?
A few lines added in developers guide:  doc/source/developer_guide/dg_dynamics.rst
- If not, does the documentation need to be updated separately at a later time?

- Other Relevant Details:

Basic structure
  evp_copyin() : gather
  evp_kernel() : loop stress/stepu/halo_update
  evp_copyout() : scatter

There is a natural "overhead" associated with gather/scatter and defining vectors. Therefore it becomes more efficient for "large" setups.

- Affected files:
./cicecore/cicedynB/dynamics/evp_kernel1d.F90 (The core, new file)
./cicecore/cicedynB/dynamics/ice_dyn_evp.F90 (switch between CICE/new kernels)
./cicecore/cicedynB/dynamics/ice_dyn_shared.F90 (namelist)
./cicecore/cicedynB/general/ice_init.F90 (namelist)
./cicecore/cicedynB/infrastructure/comm/mpi/ice_gather_scatter.F90 ()
./cicecore/cicedynB/infrastructure/comm/serial/ice_gather_scatter.F90 ()
(*) Updated gather_scatter_ext to take care of integers + logicals (icetmask,iceumask)

- Possible OMP issues in these files (see more below):
  ./cicecore/cicedynB/analysis/ice_diagnostics.F90 (2x)
  ./cicecore/cicedynB/general/ice_init.F90
  ./cicecore/drivers/cice/CICE_RunMod.F90
Update since #205 and maybe solved (to be tested):
  ./cicecore/cicedynB/analysis/ice_history.F90
  ./cicecore/cicedynB/dynamics/ice_transport_driver.F90 (2x)
  ./cicecore/cicedynB/dynamics/ice_transport_remap.F90
Update since #205: OMP -> TCXOMP
  ./cicecore/cicedynB/dynamics/ice_dyn_eap.F90 - TCXOMP
  ./cicecore/cicedynB/dynamics/ice_dyn_evp.F90 - TCXOMP

- OpenMP Issues ??:
Maybe OpenMP raises?. I have not tested it carefully, but a previous CICE version shows some raises. I have added a comment line just before the OMPs, that I did comment out last time. But all OMP's stays un-commented in this PR.
Search for "!MHRI: CHECK THIS OMP"

NOTE: I did only un-comment the OMPs to check its runs smoothly. Ie. this only means, that there is possible a thread-issue in one of the files - not necessary all of them.

mhrib commented:
Just a fast 1 node/1 core test on 1213x1228 grid (~10% ice, ~40% water). ndte=320, 2 timesteps:
Ie. almost a serial case: 

Org. CICE EVP:
Timer   1:     Total      56.59 seconds
Timer   2:  TimeLoop      42.46 seconds
Timer   3:  Dynamics      16.56 seconds
Timer   4: Advection       7.18 seconds
Timer   5:    Column      14.86 seconds
Timer   6:    Thermo      12.21 seconds
Timer   7: Shortwave       3.70 seconds
Timer   8:   Ridging       2.65 seconds
Timer   9:  Coupling       1.03 seconds
Timer  10: ReadWrite       1.21 seconds
Timer  11:     Diags       1.29 seconds
Timer  12:   History       0.14 seconds
Timer  13:     Bound       0.06 seconds
Timer  14:       BGC       0.00 seconds

EVP kernel:
Timer   1:     Total      42.76 seconds
Timer   2:  TimeLoop      28.57 seconds
Timer   3:  Dynamics       2.75 seconds
Timer   4: Advection       7.14 seconds
Timer   5:    Column      14.83 seconds
Timer   6:    Thermo      12.19 seconds
Timer   7: Shortwave       3.70 seconds
Timer   8:   Ridging       2.64 seconds
Timer   9:  Coupling       1.02 seconds
Timer  10: ReadWrite       1.27 seconds
Timer  11:     Diags       1.29 seconds
Timer  12:   History       0.14 seconds
Timer  13:     Bound       0.03 seconds
Timer  14:       BGC       0.00 seconds

NB: Dynamics also includes "copy_in" ie. forming relevant vectors from the matrices.

rgs
Mads

apcraig commented:
I agree this should not be part of the v6 release, but I think we should be able to review, test, and merge this to the trunk after the release.

mhrib commented:
Hi Elisabeth et al.,
I made the PR, as we just updated CICE to master and still want to include the new kernel. As I did that, and because there has been many updates since I made the first PR, I found it naturally to update the PR.
As it is implemented, it will not use the kernel unless evp_kernel_ver is actively changed / set to "2" within the ice_in name list.

It sounds reasonable to wait until the official release is done in order to have more eyes on the code and more tests

However, one could consider my isolated question within the PR:
Is "strocnx,strocny" used at all from subroutine "stepu"? It seems to be re-calculated in evp_finish routine and not used at all during the evp-loop.

Rgs
Mads


________________________________________
From: Elizabeth Hunke <notifications@github.com>
Sent: Tuesday, November 27, 2018 6:26:11 PM
To: CICE-Consortium/CICE
Cc: Mads Hvid Ribergaard; Mention
Subject: Re: [CICE-Consortium/CICE] Evp kernel_v2 (#252)

@eclare108213 commented on this pull request.

It looks like these changes are all isolated and shouldn't make any difference to simulations as long as the new namelist flag is set to turn them off. However we are a few days from releasing CICE v6 and I'm a little hesitant to include this much new code, which most of us don't have experience running or maintaining. I propose that we wait until after the v6 release to begin working with this new capability. I'm open to counterarguments, though, since @mhrib<https://github.com/mhrib> et al have been working with it for a while now!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/CICE-Consortium/CICE/pull/252#pullrequestreview-178912161>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AkUFxbCq4SfbKP2mNgxejU3f_yVqJWrVks5uzXWzgaJpZM4Yve0F>.

mhrib commented:
Agree, I can also update it to master when the release is official. It should be straight forward...

Rgs
Mads



________________________________________
From: Tony Craig <notifications@github.com>
Sent: Tuesday, November 27, 2018 7:08:09 PM
To: CICE-Consortium/CICE
Cc: Mads Hvid Ribergaard; Mention
Subject: Re: [CICE-Consortium/CICE] Evp kernel_v2 (#252)

I agree this should not be part of the v6 release, but I think we should be able to review, test, and merge this to the trunk after the release.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/CICE-Consortium/CICE/pull/252#issuecomment-442159499>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AkUFxSSIMTeJmf3mYw1mnG3uVEN4YKFRks5uzX-JgaJpZM4Yve0F>.

eclare108213 commented:
> However, one could consider my isolated question within the PR: Is "strocnx,strocny" used at all from subroutine "stepu"? It seems to be re-calculated in evp_finish routine and not used at all during the evp-loop.

You are correct about this.  The calculation of ice-ocean stress has changed, and now the portion that's in stepu is no longer needed.  That and the old, commented-out code in dyn_finish that went with it could be removed.  Thanks for pointing it out!

> It should be straight forward...

famous last words...

apcraig commented:
I just ran a full test suite on gordon with 4 compilers and all tests passed.  See hash 3dd936cc51 at https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks.  Timings look OK.  I updated to the master before running the tests.  A few comments/questions

- the branch should be updated to master before the PR is merged.  I can do this now if it's useful to others.  Propose rebase to not change pre-release history,  "git pull --rebase upstream master".
- should we remove the dead code (stepu strocn) in this PR or defer to another PR?  Are there any improvements that could be done with this implementation if the dead code were removed?
- What is the proper test for the new code.  I assume if I set evp_kernal_ver=2, I can check bit-for-bit, but that the code also might slow down quite a bit.  I will play around with the new feature next.
- The "stop" calls in ice_dyn_evp.F90 (line 352, 372) need to be change to abort_ice calls.  Other added stop calls, if there are any, should be modified in the same way.
- The new file, evp_kernel1d.F90 should be renamed to ice_evp_kernel1d.F90 or similar.  It should start with ice_.
- The new file, evp_kernel1d.F90 has some new cpps like _OPENMP that do not appear elsewhere in the ice code.  What do we need to do in scripts and/or documentation to make it clear how these work.  Do these cpps need to be turned on in order for the threading to work?

mhrib commented:
Hi Craig,

Comments below...

> I just ran a full test suite on gordon with 4 compilers and all tests passed. See hash 3dd936cc51 at https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks. Timings look OK. I updated to the master before running the tests. A few comments/questions
Not surprised, as the default (evp_kernel=0) is to *not* use the new kernel. If timings do change, I would expect it as "noise" between tests.

>  *   the branch should be updated to master before the PR is merged. I can do this now if it's useful to others. Propose rebase to not change pre-release history, "git pull --rebase upstream master".
I can also do it early next week - likely Monday??

>  *   should we remove the dead code (stepu strocn) in this PR or defer to another PR? Are there any improvements that could be done with this implementation if the dead code were removed?
I suggest to remove it here, as it is dead - but not computational/memory - dead. Eg. I could make a new issue on it, and you - or others - ref. to this PR. and marked it as closed as soon as the PR. has been implemented??

>  *   What is the proper test for the new code. I assume if I set evp_kernal_ver=2, I can check bit-for-bit, but that the code also might slow down quite a bit. I will play around with the new feature next.
Yes: Just set "evp_kernel=2" within the ice_in namelist and try it out. NOTE, that some OMP statements (outside the kernel) is not working correctly. 
Also note during aprun: "-d <NOMP> -cc cpu"
For intel/cray: "-d <NOMP> -cc numa_node" + "setenv KMP_AFFINITY disabled"
And of course: setenv OMP_NUM_THREADS <NOMP>

>  *   The "stop" calls in ice_dyn_evp.F90 (line 352, 372) need to be change to abort_ice calls. Other added stop calls, if there are any, should be modified in the same way.
Will do (next week)

>  *   The new file, evp_kernel1d.F90 should be renamed to ice_evp_kernel1d.F90 or similar. It should start with ice_.
Suggest to rename it to: "ice_dyn_evp_1d.F90". I will do next week

>  *   The new file, evp_kernel1d.F90 has some new cpps like _OPENMP that do not appear elsewhere in the ice code. What do we 
need to do in scripts and/or documentation to make it clear how these work. Do these cpps need to be turned on in order for the threading to work?
These flags "_OPENMP" and "_OPENACC" (gpu's) is indeed needed.
It is something the compiler itself figure out. In addition, without OpenMP/OpenACC flags during compillation (without -openmp/-omp/...), it will do a serial version.




________________________________________
From: Tony Craig <notifications@github.com>
Sent: Thursday, January 17, 2019 02:42
To: CICE-Consortium/CICE
Cc: Mads Hvid Ribergaard; Mention
Subject: Re: [CICE-Consortium/CICE] Evp kernel_v2 (#252)

I just ran a full test suite on gordon with 4 compilers and all tests passed. See hash 3dd936cc51 at https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks. Timings look OK. I updated to the master before running the tests. A few comments/questions

  *   the branch should be updated to master before the PR is merged. I can do this now if it's useful to others. Propose rebase to not change pre-release history, "git pull --rebase upstream master".
  *   should we remove the dead code (stepu strocn) in this PR or defer to another PR? Are there any improvements that could be done with this implementation if the dead code were removed?
  *   What is the proper test for the new code. I assume if I set evp_kernal_ver=2, I can check bit-for-bit, but that the code also might slow down quite a bit. I will play around with the new feature next.
  *   The "stop" calls in ice_dyn_evp.F90 (line 352, 372) need to be change to abort_ice calls. Other added stop calls, if there are any, should be modified in the same way.
  *   The new file, evp_kernel1d.F90 should be renamed to ice_evp_kernel1d.F90 or similar. It should start with ice_.
  *   The new file, evp_kernel1d.F90 has some new cpps like _OPENMP that do not appear elsewhere in the ice code. What do we need to do in scripts and/or documentation to make it clear how these work. Do these cpps need to be turned on in order for the threading to work?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/CICE-Consortium/CICE/pull/252#issuecomment-455010612>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AkUFxXMXOYmZU3VUhy2L3T-Lx6OkZBkGks5vD9ULgaJpZM4Yve0F>.

apcraig commented:
Thanks Mads, that all sounds good.  I agree an issue about removing the dead code is a good idea, and I would be happy for that code to be removed in this PR and for the issue to be associated with this PR.

I will play around with the new feature.  I am still not totally clear about the new cpps.  Does the compiler automatically set those cpps if openmp/openacc is turned on during compilation?  Is that a standard that we can rely on with all compilers/machines?  If not, I don't see other logic that would trigger the cpps automatically and we may want to address that.  I also see that it depends on omp_lib.  Is that something we can count on being on all machines and all compilers?  I suspect the answer is yes to both, just not familiar with the current standard.  I thought I've used omp_get_num_threads without a use statement before, but maybe I'm remembering wrong.

I continue to be worried about the OpenMP calls in the model.  We have only turned off ones that have created problems that we detected (like not bit-for-bit results with different pes/threads).  But apparently, there are still questions about others.  Not sure how to best move forward to validate the OpenMP.

mhrib commented:
Hi Tony,

I was also not aware of the CPPS (_OpenMP, _OpenACC) directives, but according to Jacob Weismann - as I understood him - these are "build in" in all compilers and automatically set IF the code is compiled using the compiler flag for OpenMP or OpenACC. The latter is different from compiler to compiler, but not the "_xxx" instructions. If the relevant OpenMP (OpenACC) flag is not set, then the code is reduced to a serial case. Ie. it will always work, and the serial version can be relevant to test as well. Like MPI vs. serial version.

I am also a bit worried about the OpenMP within the code for two reasons - perhaps three:
1) I am not convinced, that all the loops are thread safe. That said, I am absolutely not sure which OMP's causes problems. Perhaps the ones that I pointed out earlier within the PR, but maybe/perhaps new one have arrived since (maybe/maybe not). And likely some of the ones I pointed out is good, but I haven't checked so careful, which is a huge task to do...
2) Many of the OMPs is constructed such, that they loops over iblk exactly like it does with MPI. Ideally they should be constructed like this to be efficient (I know, that the ideal world do not exist:-)). If not, communication can take over and killing the benefit of using the nearby shared memory cache(s).

do i=1,N_MPI
...
      a=c0  (initialization of variables can be done outside OMP) 
      !$OMP PARALLEL PRIVATE(j,k,...)
          do j=1,iwet
                 Threaded loops - here within each MPI block
           enddo
      !$OMP BARRIER (needed if loops is not independent like for advection, where you use neighbour-points, which can be changed within another loop in another thread)
          do j=1,iwet
                 Threaded loops - here within each MPI block
           enddo
          Some other threaded loops
      !$OMP END PARALLEL
...
enddo

--o--

3) OpenMP is most efficient on vectors. Unfortunately matrices is an abstractation level, that is convenient, but not for compilers:-( Moreover, using vector notation it is easier to distribute the threads evenly - however harder to find neighbours for advection calculations/halos.

--o--

We - Till and I - have tried to use the CICE model in pure MPI for the entire code - expect for the evp_kernel, where we will only use OpenMP. This can be done quite simple (but ugly way) by comment-out all OMP's using a "sed"-command on every file expect for the evp_kernel file before compiling... Doing this, it is easier to compare "X MPI evp" vs "X OpenMP evp" ie. evp_kernel=0.vs.2 using the same number of cores. And scaling as well using eg. ndte=1200 iterations. Note, that you will relatively quickly be limited by the "2d_to_1d" and "1d_to_2d" transforms. At least for our 1213x1228 point case. Therefore Till have added a new timer in ice_dyn_evp.F90 like this to get an idea of the potential of vectorize the entire code (a huge job!)

         call ice_timer_start(timer_evp_1d)
          call evp_kernel_v2()
          call ice_timer_stop(timer_evp_1d)

rgs
Mads


________________________________________
From: Tony Craig <notifications@github.com>
Sent: Thursday, January 17, 2019 18:06
To: CICE-Consortium/CICE
Cc: Mads Hvid Ribergaard; Mention
Subject: Re: [CICE-Consortium/CICE] Evp kernel_v2 (#252)

Thanks Mads, that all sounds good. I agree an issue about removing the dead code is a good idea, and I would be happy for that code to be removed in this PR and for the issue to be associated with this PR.

I will play around with the new feature. I am still not totally clear about the new cpps. Does the compiler automatically set those cpps if openmp/openacc is turned on during compilation? Is that a standard that we can rely on with all compilers/machines? If not, I don't see other logic that would trigger the cpps automatically and we may want to address that. I also see that it depends on omp_lib. Is that something we can count on being on all machines and all compilers? I suspect the answer is yes to both, just not familiar with the current standard. I thought I've used omp_get_num_threads without a use statement before, but maybe I'm remembering wrong.

I continue to be worried about the OpenMP calls in the model. We have only turned off ones that have created problems that we detected (like not bit-for-bit results with different pes/threads). But apparently, there are still questions about others. Not sure how to best move forward to validate the OpenMP.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/CICE-Consortium/CICE/pull/252#issuecomment-455250403>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AkUFxSn2Ir-zVTUrTX9poIwFnx8087S1ks5vEK2AgaJpZM4Yve0F>.

mhrib commented:
I have now updated the EVP kernel to master.

It is push'ed to here:
https://github.com/mhrib/CICE/tree/EVP_kernelv2

Forgot how to push it from there to here:
https://github.com/CICE-Consortium/CICE/pull/252

rgs
Mads

________________________________________
From: Tony Craig <notifications@github.com>
Sent: Thursday, January 17, 2019 18:06
To: CICE-Consortium/CICE
Cc: Mads Hvid Ribergaard; Mention
Subject: Re: [CICE-Consortium/CICE] Evp kernel_v2 (#252)

Thanks Mads, that all sounds good. I agree an issue about removing the dead code is a good idea, and I would be happy for that code to be removed in this PR and for the issue to be associated with this PR.

I will play around with the new feature. I am still not totally clear about the new cpps. Does the compiler automatically set those cpps if openmp/openacc is turned on during compilation? Is that a standard that we can rely on with all compilers/machines? If not, I don't see other logic that would trigger the cpps automatically and we may want to address that. I also see that it depends on omp_lib. Is that something we can count on being on all machines and all compilers? I suspect the answer is yes to both, just not familiar with the current standard. I thought I've used omp_get_num_threads without a use statement before, but maybe I'm remembering wrong.

I continue to be worried about the OpenMP calls in the model. We have only turned off ones that have created problems that we detected (like not bit-for-bit results with different pes/threads). But apparently, there are still questions about others. Not sure how to best move forward to validate the OpenMP.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/CICE-Consortium/CICE/pull/252#issuecomment-455250403>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AkUFxSn2Ir-zVTUrTX9poIwFnx8087S1ks5vEK2AgaJpZM4Yve0F>.

apcraig commented:
I am running some tests now.

We have another problem in that the update to master was not done as a rebase so all the master changes are appearing in this commit.  I think we believe it will not affect the PR merge but it does make it difficult to review.  We might want to consider making a new branch off the consortium master and then merging in this branch on top.  That's what a rebase pull would have done.  For now, I'll proceed with testing though.

apcraig commented:
Test results are here, https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks, hash aa6de33.  Looks like some box configurations changed answers for intel, pgi, and gnu but not cray.  Otherwise everything is ok.  Do we need to understand why some of the box configurations have different answers?

mhrib commented:
Hi Tony,

I did "git pull --rebase upstream master" and afterwards changed a few things in one file, that did not merge out-of-the-box.

Should I do it in another fashion, or make a new PR?

I am absolutely not an expert in git, but if you could give me some hints I may be able to fix it so it will keep the comments etc?

Rgs
Mads



________________________________________
From: Tony Craig <notifications@github.com>
Sent: Sunday, January 27, 2019 1:52:02 AM
To: CICE-Consortium/CICE
Cc: Mads Hvid Ribergaard; Mention
Subject: Re: [CICE-Consortium/CICE] Evp kernel_v2 (#252)

I am running some tests now.

We have another problem in that the update to master was not done as a rebase so all the master changes are appearing in this commit. I think we believe it will not affect the PR merge but it does make it difficult to review. We might want to consider making a new branch off the consortium master and then merging in this branch on top. That's what a rebase pull would have done. For now, I'll proceed with testing though.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/CICE-Consortium/CICE/pull/252#issuecomment-457879234>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AkUFxWD2QVBjZoFlHDCnk86nkgKS7Be3ks5vHPgygaJpZM4Yve0F>.

apcraig commented:
@mhrib I look a bit closer.  You are correct, it looks like you did a rebase based on the log on your branch.  I expected this would result in a cleaner representation of the files changed in the PR, but it doesn't.  That's a bit frustrating.  Lets leave this branch as is for the moment.

mhrib commented:
@apcraig: Note, that there is also changes in mpi/serial scatter_global and gather_global_ext making them able to handle integers and logicals for passing /icetmask, iceumask) matrices. This is quite some lines - however very much similar to the existing gather/scatter routines. The gather_global_ext has been transformed into an interface.



________________________________________
From: Tony Craig <notifications@github.com>
Sent: Sunday, January 27, 2019 20:48
To: CICE-Consortium/CICE
Cc: Mads Hvid Ribergaard; Mention
Subject: Re: [CICE-Consortium/CICE] Evp kernel_v2 (#252)

@mhrib<https://github.com/mhrib> I look a bit closer. You are correct, it looks like you did a rebase based on the log on your branch. I expected this would result in a cleaner representation of the files changed in the PR, but it doesn't. That's a bit frustrating. Lets leave this branch as is for the moment.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/CICE-Consortium/CICE/pull/252#issuecomment-457948114>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AkUFxWNnFplTaQe1-HXO0J9GgPXe_C4gks5vHgJ3gaJpZM4Yve0F>.

mhrib commented:
@apcraig: Not sure I understand the differences of your test runs. I assume all did compile, but is it different box-decompositions that fails for some compilers and not for others?
If so, the errors could also stems from OMP problems within other routines. This can be checked "simply" (sed command) by comment-out all OMP statements in all routines except for the ones in ice_dyn_evp_1d.F90

________________________________________
From: Tony Craig <notifications@github.com>
Sent: Sunday, January 27, 2019 05:11
To: CICE-Consortium/CICE
Cc: Mads Hvid Ribergaard; Mention
Subject: Re: [CICE-Consortium/CICE] Evp kernel_v2 (#252)

Test results are here, https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks, hash aa6de33<https://github.com/CICE-Consortium/CICE/commit/aa6de33f198ce2952cbd43956265bed8890a1fa2>. Looks like some box configurations changed answers for intel, pgi, and gnu but not cray. Otherwise everything is ok. Do we need to understand why some of the box configurations have different answers?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/CICE-Consortium/CICE/pull/252#issuecomment-457887549>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AkUFxRIHW3lOXspS-8HN7cFWUdYDUf4Aks5vHSbYgaJpZM4Yve0F>.

apcraig commented:
Getting back to the box test failures.  The model runs fine, the problem is the answer is different for several (but not all) box cases compared to the current master.  This PR is bit-for-bit for all non-box cases with all compilers.  I will look a bit further.   But I doubt it's openmp as that is not likely to affect just the box cases and not others.  I think the thing that jumps out at me is the boundary conditions.  The box case has different boundary conditions than the global grids, at least in the "longitudinal" direction where non-box cases wrap around.  Could there be a difference/error in the implementation for cases with non-wrap-around boundary conditions?  I will try to look into this a bit more.  @mhrib Is it possible for you to run some of the cases.  You can see all the test results here, https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks under hash aa6de33.  You can see that 5 tests failed bit-for-bit comparison on each pgi, intel, gnu.  But they all passed on cray.  That's also a bit odd.  It could be a compiler/omp thing or something else.  Is the omp behaving differently in this code compared to the master.  We test all mpi, openmp, and mixed pretty carefully and compare answers between them all when we can.  That is how we have detected other openmp problems in the past.  I'm not saying there aren't problems, but I just don't understand why it only affects a few of the box cases and only on 3/4 compilers on gordon.

apcraig commented:
I found a good failing test.  I cut below a case where openmp is off, we are running the gbox128 on 8x1 tasks x threads, diagnostic written every timestep.

Current master:

```
istep1:         1    idate:  19970101    sec:      3600
  Initial forcing data year =         1997
  Final   forcing data year =         1997
  Current forcing data year =         1997
  
 Finished writing ./history/iceh_ic.1997-01-01-00000.nc
                                             Arctic                 Antarctic
total ice area  (km^2) =    7.14237020674198773E+06   0.00000000000000000E+00
total ice extent(km^2) =    1.21635000000000000E+07   0.00000000000000000E+00
total ice volume (m^3) =    1.42960388367532070E+13   0.00000000000000000E+00
total snw volume (m^3) =    1.42853391809006396E+12   0.00000000000000000E+00
tot kinetic energy (J) =    1.04308509283625344E+14   0.00000000000000000E+00
rms ice speed    (m/s) =        0.12393981696376018       0.00000000000000000
average albedo         =        1.00000000000000000       0.00000000000000000
max ice volume     (m) =        1.99371521192846801  ************************
max ice speed    (m/s) =        0.17140107227822357  ************************
max strength    (kN/m) =      155.95968782447957324  ************************
 ----------------------------
arwt rain h2o kg in dt =    0.00000000000000000E+00   0.00000000000000000E+00
arwt snow h2o kg in dt =    0.00000000000000000E+00   0.00000000000000000E+00
arwt evap h2o kg in dt =    1.91513434279816513E+10   0.00000000000000000E+00
arwt frzl h2o kg in dt =    0.00000000000000000E+00   0.00000000000000000E+00
arwt frsh h2o kg in dt =   -1.03046549289842930E+13   0.00000000000000000E+00
arwt ice mass (kg)     =    1.31094676133026900E+16   0.00000000000000000E+00
arwt snw mass (kg)     =    4.71416192969721125E+14   0.00000000000000000E+00
arwt tot mass (kg)     =    1.35808838062724120E+16   0.00000000000000000E+00
arwt tot mass chng(kg) =    1.03238062724120000E+13   0.00000000000000000E+00
arwt water flux        =    1.03238062724122754E+13   0.00000000000000000E+00
 (=rain+snow+evap+frzl-fresh)  
water flux error       =    2.02778132062958380E-17   0.00000000000000000E+00
 ----------------------------
arwt atm heat flux (W) =   -5.14185247756211328E+13   0.00000000000000000E+00
arwt ocn heat flux (W) =   -1.17835525201700000E+14   0.00000000000000000E+00
arwt frzl heat flux(W) =    0.00000000000000000E+00   0.00000000000000000E+00
arwt tot energy    (J) =   -4.76571858097386029E+21   0.00000000000000000E+00
arwt net heat      (J) =    2.39101201533883936E+17   0.00000000000000000E+00
arwt tot energy chng(J)=    2.39101201534156800E+17   0.00000000000000000E+00
arwt heat error        =    5.72555838880945994E-17   0.00000000000000000E+00
 ----------------------------
arwt incoming sw (W)   =    0.00000000000000000E+00   0.00000000000000000E+00
arwt absorbed sw (W)   =    0.00000000000000000E+00   0.00000000000000000E+00
arwt swdn error        =    0.00000000000000000E+00   0.00000000000000000E+00
 ----------------------------
total brine tr (m^3)   =    0.00000000000000000E+00   0.00000000000000000E+00
arwt salt mass (kg)    =    5.24378704532107578E+13   0.00000000000000000E+00
arwt salt mass chng(kg)=    4.12240532107600021E+10   0.00000000000000000E+00
arwt salt flx in dt(kg)=   -4.12240532107647934E+10   0.00000000000000000E+00
arwt salt flx error    =   -9.13702201141072666E-17   0.00000000000000000E+00
 ----------------------------
                          
       Lat, Long                     89.9   23.6           71.8 -122.4
  my_task, iblk, i, j            4    1    3    8          3    1   33    4
 ----------atm----------
air temperature (C)    =      -20.14999999999997726     -20.14999999999997726
specific humidity      =        0.00060000000000000       0.00060000000000000
snowfall (m)           =        0.00000000000000000       0.00000000000000000
rainfall (m)           =        0.00000000000000000       0.00000000000000000
shortwave radiation sum=        0.00000000000000000       0.00000000000000000
longwave radiation     =      180.00000000000000000     180.00000000000000000
 ----------ice----------
area fraction          =        0.01170127822002812       0.99160553732165613
avg ice thickness (m)  =        2.00157353786122405       2.00168379704297283
avg snow depth (m)     =        0.20000772644486051       0.20001128813199259
avg salinity (ppt)     =        2.32551636526343097       2.32551636526344252
avg brine thickness (m)=        0.00000000000000000       0.00000000000000000
surface temperature(C) =      -22.28674009109586862     -22.34808630976561261
absorbed shortwave flx =        0.00000000000000000       0.00000000000000000
outward longwave flx   =     -222.66327458031167907    -223.12825998110193382
sensible heat flx      =       34.12714109554898556      33.80801285663843458
latent heat flx        =        2.04113905604011636       2.08830094793734666
subl/cond (m ice)      =        0.00000278470764555       0.00000284904149604
top melt (m)           =        0.00000000000000000       0.00000000000000000
bottom melt (m)        =        0.00000000000000000       0.00000000000000000
lateral melt (m)       =        0.00000000000000000       0.00000000000000000
new ice (m)            =        0.00000000000000000       0.00000000000000000
congelation (m)        =        0.00001843989680427       0.00156739122836325
snow-ice (m)           =        0.00000000000000000       0.00000000000000000
snow change (m)        =        0.00000000000000000       0.00000000000000000
effective dhi (m)      =       -0.00001653115563983      -0.00730676288514998
effective dhs (m)      =       -0.00000340394671341      -0.00088644916147890
intnl enrgy chng(W/m^2)=       -3.35542558169561556    -831.78274165365428416
 ----------ocn----------
sst (C)                =       -1.90458264992426463      -1.90458264992426463
sss (ppt)              =       34.00000000000000000      34.00000000000000000
freezing temp (C)      =       -1.90458264992426463      -1.90458264992426463
heat used (W/m^2)      =       16.50273655272619777      16.81543837335271618

```
Current version:

```
istep1:         1    idate:  19970101    sec:      3600
  Initial forcing data year =         1997
  Final   forcing data year =         1997
  Current forcing data year =         1997
  
 Finished writing ./history/iceh_ic.1997-01-01-00000.nc
                                             Arctic                 Antarctic
total ice area  (km^2) =    7.14237020674198866E+06   0.00000000000000000E+00
total ice extent(km^2) =    1.21635000000000000E+07   0.00000000000000000E+00
total ice volume (m^3) =    1.42960388367532070E+13   0.00000000000000000E+00
total snw volume (m^3) =    1.42853391809006396E+12   0.00000000000000000E+00
tot kinetic energy (J) =    1.04308509283625297E+14   0.00000000000000000E+00
rms ice speed    (m/s) =        0.12393981696376015       0.00000000000000000
average albedo         =        0.99999999999999989       0.00000000000000000
max ice volume     (m) =        1.99371521192844514  ************************
max ice speed    (m/s) =        0.17140107227822357  ************************
max strength    (kN/m) =      155.95968782447957324  ************************
 ----------------------------
arwt rain h2o kg in dt =    0.00000000000000000E+00   0.00000000000000000E+00
arwt snow h2o kg in dt =    0.00000000000000000E+00   0.00000000000000000E+00
arwt evap h2o kg in dt =    1.91513434279816513E+10   0.00000000000000000E+00
arwt frzl h2o kg in dt =    0.00000000000000000E+00   0.00000000000000000E+00
arwt frsh h2o kg in dt =   -1.03046549289842930E+13   0.00000000000000000E+00
arwt ice mass (kg)     =    1.31094676133026900E+16   0.00000000000000000E+00
arwt snw mass (kg)     =    4.71416192969721125E+14   0.00000000000000000E+00
arwt tot mass (kg)     =    1.35808838062724120E+16   0.00000000000000000E+00
arwt tot mass chng(kg) =    1.03238062724120000E+13   0.00000000000000000E+00
arwt water flux        =    1.03238062724122754E+13   0.00000000000000000E+00
 (=rain+snow+evap+frzl-fresh)  
water flux error       =    2.02778132062958380E-17   0.00000000000000000E+00
 ----------------------------
arwt atm heat flux (W) =   -5.14185247756211328E+13   0.00000000000000000E+00
arwt ocn heat flux (W) =   -1.17835525201699937E+14   0.00000000000000000E+00
arwt frzl heat flux(W) =    0.00000000000000000E+00   0.00000000000000000E+00
arwt tot energy    (J) =   -4.76571858097385924E+21   0.00000000000000000E+00
arwt net heat      (J) =    2.39101201533883712E+17   0.00000000000000000E+00
arwt tot energy chng(J)=    2.39101201535205376E+17   0.00000000000000000E+00
arwt heat error        =    2.77327327987109579E-16   0.00000000000000000E+00
 ----------------------------
arwt incoming sw (W)   =    0.00000000000000000E+00   0.00000000000000000E+00
arwt absorbed sw (W)   =    0.00000000000000000E+00   0.00000000000000000E+00
arwt swdn error        =    0.00000000000000000E+00   0.00000000000000000E+00
 ----------------------------
total brine tr (m^3)   =    0.00000000000000000E+00   0.00000000000000000E+00
arwt salt mass (kg)    =    5.24378704532107578E+13   0.00000000000000000E+00
arwt salt mass chng(kg)=    4.12240532107600021E+10   0.00000000000000000E+00
arwt salt flx in dt(kg)=   -4.12240532107647934E+10   0.00000000000000000E+00
arwt salt flx error    =   -9.13702201141072666E-17   0.00000000000000000E+00
 ----------------------------
                          
       Lat, Long                     89.9   23.6           71.8 -122.4
  my_task, iblk, i, j            4    1    3    8          3    1   33    4
 ----------atm----------
air temperature (C)    =      -20.14999999999997726     -20.14999999999997726
specific humidity      =        0.00060000000000000       0.00060000000000000
snowfall (m)           =        0.00000000000000000       0.00000000000000000
rainfall (m)           =        0.00000000000000000       0.00000000000000000
shortwave radiation sum=        0.00000000000000000       0.00000000000000000
longwave radiation     =      180.00000000000000000     180.00000000000000000
 ----------ice----------
area fraction          =        0.01170127822002812       0.99160553732165624
avg ice thickness (m)  =        2.00157353786122405       2.00168379704297283
avg snow depth (m)     =        0.20000772644486051       0.20001128813199259
avg salinity (ppt)     =        2.32551636526343097       2.32551636526344252
avg brine thickness (m)=        0.00000000000000000       0.00000000000000000
surface temperature(C) =      -22.28674009109586862     -22.34808630976561261
absorbed shortwave flx =        0.00000000000000000       0.00000000000000000
outward longwave flx   =     -222.66327458031167907    -223.12825998110190540
sensible heat flx      =       34.12714109554898556      33.80801285663842748
latent heat flx        =        2.04113905604011636       2.08830094793734622
subl/cond (m ice)      =        0.00000278470764555       0.00000284904149604
top melt (m)           =        0.00000000000000000       0.00000000000000000
bottom melt (m)        =        0.00000000000000000       0.00000000000000000
lateral melt (m)       =        0.00000000000000000       0.00000000000000000
new ice (m)            =        0.00000000000000000       0.00000000000000000
congelation (m)        =        0.00001843989680427       0.00156739122836325
snow-ice (m)           =        0.00000000000000000       0.00000000000000000
snow change (m)        =        0.00000000000000000       0.00000000000000000
effective dhi (m)      =       -0.00001653115563983      -0.00730676288514975
effective dhs (m)      =       -0.00000340394671341      -0.00088644916147887
intnl enrgy chng(W/m^2)=       -3.35542558169561556    -831.78274165358811842
 ----------ocn----------
sst (C)                =       -1.90458264992426463      -1.90458264992426463
sss (ppt)              =       34.00000000000000000      34.00000000000000000
freezing temp (C)      =       -1.90458264992426463      -1.90458264992426463
heat used (W/m^2)      =       16.50273655272619777      16.81543837335268776

```

```
The diffs are

< total ice area  (km^2) =    7.14237020674198773E+06   0.00000000000000000E+00
---
> total ice area  (km^2) =    7.14237020674198866E+06   0.00000000000000000E+00
354,357c355,358
< tot kinetic energy (J) =    1.04308509283625344E+14   0.00000000000000000E+00
< rms ice speed    (m/s) =        0.12393981696376018       0.00000000000000000
< average albedo         =        1.00000000000000000       0.00000000000000000
< max ice volume     (m) =        1.99371521192846801  ************************
---
> tot kinetic energy (J) =    1.04308509283625297E+14   0.00000000000000000E+00
> rms ice speed    (m/s) =        0.12393981696376015       0.00000000000000000
> average albedo         =        0.99999999999999989       0.00000000000000000
> max ice volume     (m) =        1.99371521192844514  ************************
375c376
< arwt ocn heat flux (W) =   -1.17835525201700000E+14   0.00000000000000000E+00
---
> arwt ocn heat flux (W) =   -1.17835525201699937E+14   0.00000000000000000E+00
377,380c378,381
< arwt tot energy    (J) =   -4.76571858097386029E+21   0.00000000000000000E+00
< arwt net heat      (J) =    2.39101201533883936E+17   0.00000000000000000E+00
< arwt tot energy chng(J)=    2.39101201534156800E+17   0.00000000000000000E+00
< arwt heat error        =    5.72555838880945994E-17   0.00000000000000000E+00
---
> arwt tot energy    (J) =   -4.76571858097385924E+21   0.00000000000000000E+00
> arwt net heat      (J) =    2.39101201533883712E+17   0.00000000000000000E+00
> arwt tot energy chng(J)=    2.39101201535205376E+17   0.00000000000000000E+00
> arwt heat error        =    2.77327327987109579E-16   0.00000000000000000E+00
403c404
< area fraction          =        0.01170127822002812       0.99160553732165613
---
> area fraction          =        0.01170127822002812       0.99160553732165624
410,412c411,413
< outward longwave flx   =     -222.66327458031167907    -223.12825998110193382
< sensible heat flx      =       34.12714109554898556      33.80801285663843458
< latent heat flx        =        2.04113905604011636       2.08830094793734666
---
> outward longwave flx   =     -222.66327458031167907    -223.12825998110190540
> sensible heat flx      =       34.12714109554898556      33.80801285663842748
> latent heat flx        =        2.04113905604011636       2.08830094793734622
421,423c422,424
< effective dhi (m)      =       -0.00001653115563983      -0.00730676288514998
< effective dhs (m)      =       -0.00000340394671341      -0.00088644916147890
< intnl enrgy chng(W/m^2)=       -3.35542558169561556    -831.78274165365428416
---
> effective dhi (m)      =       -0.00001653115563983      -0.00730676288514975
> effective dhs (m)      =       -0.00000340394671341      -0.00088644916147887
> intnl enrgy chng(W/m^2)=       -3.35542558169561556    -831.78274165358811842
428c429
< heat used (W/m^2)      =       16.50273655272619777      16.81543837335271618
---
> heat used (W/m^2)      =       16.50273655272619777      16.81543837335268776

``` 
So this suggests it's not openmp.  It also suggests the initial divergence is roundoff.  If we beleive these changes might be introducing a roundoff error (order or operations or similar), then maybe this is all fine.  I still wonder why this only shows up on some box cases and not other cases.

apcraig commented:
Let me summarize where we are.  

With evp_kernel_ver=0, results are bit-for-bit for most tests against the current master.  This is running full test suites on gordon for 4 compilers.  A subset of box tests are NOT bit-for-bit on 3/4 compilers.  Rerunning the failed box tests with the debug flag (reduced optimization and run time checks) on both master and this PR results in bit-for-bit identical answers.  It seems the changes in the answers in the box test is caused by some compiler optimization as a results of the code changes.  This might be associated with the evp kernel changes (although @mhrib makes a case it shouldn't) or it might be associated with some of the code cleanup.  We could look into this further or we could accept it.  Personally, I am comfortable with this outcome as it stands.  I believe we've shown the answers are roundoff different (see above gbox128 diff) as a result of compiler optimization and that we can make this bit-for-bit if we reduce compiler optimization.  I think based on these results, we could merge this PR.  evp_kernel_ver=0 will be the default setting.

Separately, there is an effort to test and validate the evp_kernel_ver=2.  The same test suite on gordon was run with the new kernel on.  Results can be found https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks, hash aa6de33...+evpk=2.   Three to four tests fail on each compiler, and they are the same tests across the compilers.  Looking at the intel results, https://github.com/CICE-Consortium/Test-Results/wiki/aa6de33f19.gordon.pgi.190128.235649, there are four failures.

- restart gbox128 4x2.  This test runs but fails to restart exactly. 
- restart gx1 40x4 droundrobin medium.  This test fails with "(abort_ice) error = (horizontal_remap)ERROR: bad departure points" on the first timestep.
- restart gx3 16x2x5x10x20 drakeX2.  This test fails with "(abort_ice) error = (horizontal_remap)ERROR: bad departure points" on the first timestep.
- restart tx1 40x4 dsectrobin medium.  This test fails gracefully in the evp kernel.  tx1 is not supported yet.

Again, many tests passed, but these 4 failures need to be debugged.  In addition, the qc test relies on the gx1 configuration, so the qc testing comparing evp_kernel_ver=2 to 0 could not be done.

So, the outstanding tasks are

- debug the 4 failures noted above
- run the qc test comparing evp_kernel_ver=0 to evp_kernel_ver=2.  This requires gx1 (one of the failing tests)
- update documentation
- produce and document timing information comparing evp_kernel_ver=0 and 2.
- add evp_kernel_ver=2 tests to the test suite
- maybe do a little cleanup on ice_dyn_evp_1d.F90 to make the code a little more readable (breaks between subroutines and such)

I have reviewed the code diffs by doing a diff in two sandboxes (the file diffs in the PR contain a bunch of changes from the master merge).  I am fine with the current code.  And I am comfortable with the evp_kernel_ver=0 implementation.  I propose we merge this PR now with the understanding that we'd create a new issue that identified the outstanding tasks above still to be done.  @eclare108213, are you OK with that approach.

apcraig commented:
This was merged as #278, exact same code changes just on a clean branch.
